### ISSUE TITLE
Add fixed ID to bounding box in Cvat Xml

### ISSUE THREAD

**Issue by egrinstein** [opencv/cvat#893]
Hello,

We are trying to show annotators how we would like our objects annotated by sending them a .pdf report. This report referrences every frame, every bounding box and how we would like for things to be different in it.

The problem is that every time a box is added, all the ids of all boxes of all frames change as well. 
Even within a frame, the id order changes.

What I think should be done in the code is:
- Create a counter which increases every time a bounding box is created.
- Assign the current value of the counter to every new bounding box created
- When loading/dumping an annotation, use the attribute "id" inside <box> as id.

The changes in the .xml should be:
- Create a tag under <task> called <id_counter> holding the next id value
- Create an attribute under <box> called "id"

What do you think, @nmanovic ? Does that make sense? Would that break something in the interpolation mode or something?

Thank you very much,

Eric

**Comment by nmanovic** [comment]
@egrinstein,

It is a complex topic. There are many problems here. Previously we had a client id but got a lot of problems. 

1. Which code should assign ids? Client? What if several clients edit the same task?
2. CVAT server cannot trust clients to assign ids by default. Need to check/validate and it is performance issues.
3. If we find an issue with assigned ids on the server what should we do? Reject the request? Indeed a user can open a couple of tabs on the same machine and it will give us a problem with saving annotations. We had the pain.
4. If we assign all ids on the server it will require constant synchronization between the server and its clients. I'm not sure that it is good. Need to think more about the option. 

P.S. I don't want to say that it is impossible to implement. It is not so trivial as it looks.

**Comment by nmanovic** [comment]
I will close the issue for now. Don't hesitate to reopen it if you have any ideas how to implement the feature.

**Comment by EvgenyM** [comment]
@nmanovic As I can see, a server Id is present and is used to generate object URL. Example: `https://cvat.xyz.com/?id=58&frame=0&filter=*%5BserverID%3D%2267145%22%5D`

Having this id in the .dump files would tremendously help error detection and QA process. What do you think?

**Comment by nmanovic** [comment]
@EvgenyM , I think it can. The main problem here that the ID can be changed after some time. For example, let's say you update a shape with `server id #X`. Sometimes on server it is more efficient to delete the object and recreate it from scratch (with another `server id #Y`). Thus the previous dump will not be valid even the shape is mostly the same.
Then we start to think about sync ids between client and server it becomes big pain. Every time when you create an object on client you have to go to the server and allocate an ID (it is a performance issue). Previously the client generated an ID but it led to many problems especially when people tried to change the same job in parallel.

Could you please submit another issue about simplifying validation process? Describe your use case. Probably we will find an alternative solution.

**Comment by ehofesmann** [comment]
> The main problem here that the ID can be changed after some time. For example, let's say you update a shape with `server id #X`. Sometimes on server it is more efficient to delete the object and recreate it from scratch (with another `server id #Y`). Thus the previous dump will not be valid even the shape is mostly the same. Then we start to think about sync ids between client and server it becomes big pain. Every time when you create an object on client you have to go to the server and allocate an ID (it is a performance issue). Previously the client generated an ID but it led to many problems especially when people tried to change the same job in parallel.

@nmanovic Can you provide some insight into what operations might cause the object to be deleted and recreated resulting in the `server id` changing? (Assuming this behavior still exists since the time you wrote this)

Consistency is a fairly important property for object IDs. In my case, I am uploading annotations and trying to keep track of which objects have been modified/added/deleted which requires a mapping between my custom object id and the CVAT object's `server id`. The fact that the `server id` can change would break this mapping and result in unwanted behavior. (Any other ideas of how to link CVAT objects with external object ids in a stable way would be welcome as well!)

A possible solution for your case could be to switch to using UUIDs instead of integer ids as this would allow clients to create ids without the concern of any ids overlapping. I assume this would require a significant code overhaul, though...

**Comment by nmanovic** [comment]
@ehofesmann , it is better to look at code to get the answer. But probably any update operation will lead to the behavior. Probably the feature which you have mentioned is very important. Don't hesitate to create a new issue and describe your needs. We will see how to support. Also we are happy to accept any contribution in the direction. There is bulk_update method in Django. Probably we can use it to avoid "deleting" objects.
